### PR TITLE
RFC: create a list of device pointers per driver subsystem

### PR DESCRIPTION
**Pull request by galak**
This the starts of a means to allow a list of devices per driver
class/subsystem for things like the shell to utilize.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

### CODE REVIEW

**Comment by galak** [comment]
The idea here is to help replace the need for things like DT_ADC_{0..n}_NAME that get used by the shell to know which devices are ADCs.  The problem with the scheme we've used is that it requires setting the magic #defines somewhere.  It also doesn't deal with the case of non-SoC devices well.  Additionally something like the I2C shell makes assumptions of "names" for I2C being prefixed with "I2C_" which isn't always true.

So the idea is we add some kinda of REGISTER macro that will create a list of device pointers per driver subsystem.  We can then utilize that list in the shell or tests as we need.

The PR is mostly to start the discussion, as well as bike shed the names of things.

File: drivers/adc/adc_mcux_adc16.c [hunk 1]
```diff
@@ -305,3 +305,13 @@ static const struct adc_driver_api mcux_adc16_driver_api = {
 	}
 
 DT_INST_FOREACH(ACD16_MCUX_INIT)
+
+#define SUBSYS_DEVICE_REGISTER(subsys, name) \
```
Comment (tbursztyka) [inline drivers/adc/adc_mcux_adc16.c:309]: I would see such registration being related to the shell, so it would not register devices (and thus take ROM) if nothing is using this. (if we get other use case, we'll enable the registration for these as well later on)
if device shell module is enabled: all devices would be registered.
If only i2c shell module is enabled, only i2c devices would be registered. (same for sensors etc...)
i2c registration would check whether device or i2c shell is enabled.

That being said:
- I would not follow that path _unless_ we get rid of the device names. Let's not add a feature replacing (and improving) an existing one, if we cannot remove the old one.
- If we do such registration, let's make it simple: DEVICE_DEFINE() would take the subsys parameter. (Note that I am talking about the right DEVICE_DEFINE we should have someday: that will be unique, that will transparently handle PM, and which would not have competitors such as DEVICE_AND_API_INIT which one would not exist anymore)

**Review by pabigot — COMMENTED** [review]
I can see the use case, but I'm not convinced this is the right approach.

First, I would like to see this integrated with #22941.  I feel that a better general solution to this sort of problem is to record the associated subsystem and other attributes in the const driver structure and allow the list to be built up by walking the device list identifying devices that meet specific conditions.  This is also IMO more forward-looking, as devices may be enabled or disabled in the future, which would also be recorded in the (non-const) driver state and would be important for the shell to know.  (In fact the device subsystem could be derived from the api pointer, no additional storage required.)

If we do decide it has to be done by recording a set of devices, I would prefer to see that use a more efficent representation.  I've been pushing for integral (16-bit) handles as the canonical device identifier, with special values for undefined, unknown, and an end-of-sequence marker.  My concept for this is that the const device structures are generated in an array (as with the init records today, but without explicit order), including linker symbols that correspond to the offset of the device within the array.  Then `DT_DEVICE_HANDLE(node)` would produce an external symbol which would resolve at link time to a constant index for the node; `DT_DEVICE(node)` would then be `__device_start + DT_DEVICE_HANDLE(node)`.

In short I think it's too early to be picking solutions, though certainly not too early to discuss and plan for them.

**Comment by tbursztyka** [comment]
> If we do decide it has to be done by recording a set of devices, I would prefer to see that use a more efficent representation. I've been pushing for integral (16-bit) handles as the canonical device identifier,

Good point, I proposed something similar for the PM dependency tree in  #22941 but it could be more widely used. Also, on a 100% DTS based build, the size size could be reduce to 8bits.
